### ISSUE TITLE
Abort due to double-panic

### ISSUE THREAD

**Issue by Michael-F-Bryan** [colin-kiegel/rust-pretty-assertions#96]
When you use the `assert_eq!()` macro, any panics that are triggered as part of formatting the diff will abort the process. Due to `cargo test`'s default behaviour, the original panic message will also be swallowed unless you use the `--nocapture` flag.

In particular, I ran into this `panic!()`:

https://github.com/colin-kiegel/rust-pretty-assertions/blob/9654759b57a88cec10fd4ffac563a93a80006689/pretty_assertions/src/printer.rs#L103-L106

Which is called while printing the `Comparison` during the failed assertion.

https://github.com/colin-kiegel/rust-pretty-assertions/blob/9654759b57a88cec10fd4ffac563a93a80006689/pretty_assertions/src/lib.rs#L140

<details>
<summary>Full Backtrace</summary>

```
$ cargo t -- --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.07s
     Running unittests src/lib.rs (/home/consulting/Documents/hotg-ai/wit-bindgen-wasm3/target/debug/deps/wit_bindgen_gen_wasm3-faa01879e61ccf80)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/codegen.rs (/home/consulting/Documents/hotg-ai/wit-bindgen-wasm3/target/debug/deps/codegen-e5f76d6e17dfe8c2)

running 1 test
thread 'single_function' panicked at 'insertion followed by deletion', /home/consulting/.cargo/registry/src/github.com-1ecc6299db9ec823/pretty_assertions-1.2.0/src/printer.rs:105:17
stack backtrace:
   0:     0x560e52b89a4c - std::backtrace_rs::backtrace::libunwind::trace::h3378927f541f6c04
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x560e52b89a4c - std::backtrace_rs::backtrace::trace_unsynchronized::hdb4c2fa83db93444
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x560e52b89a4c - std::sys_common::backtrace::_print_fmt::hcd86ae48a4354c57
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x560e52b89a4c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h39278ceede34f184
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x560e52baf38c - core::fmt::write::hdc75737b203416ad
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/fmt/mod.rs:1190:17
   5:     0x560e52b84a18 - std::io::Write::write_fmt::hd5a5b2845cf361c8
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/io/mod.rs:1657:15
   6:     0x560e52b8bd17 - std::sys_common::backtrace::_print::h56e98f99dedad7ea
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x560e52b8bd17 - std::sys_common::backtrace::print::hac723d22fa420755
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x560e52b8bd17 - std::panicking::default_hook::{{closure}}::h848d7a113757d5f2
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:295:22
   9:     0x560e52b8b9cf - std::panicking::default_hook::hfb349cb28b3534e6
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:314:9
  10:     0x560e52b8c46b - std::panicking::rust_panic_with_hook::h0a743d387214ad02
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:698:17
  11:     0x560e52b33248 - std::panicking::begin_panic::{{closure}}::hfca5b139e785ec6a
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:617:9
  12:     0x560e52b31e09 - std::sys_common::backtrace::__rust_end_short_backtrace::h469676dac8d9d04d
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x560e52b331dc - std::panicking::begin_panic::ha6796eda940863d0
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:616:12
  14:     0x560e52955aee - pretty_assertions::printer::write_lines::hdf106cf409ad0dd8
                               at /home/consulting/.cargo/registry/src/github.com-1ecc6299db9ec823/pretty_assertions-1.2.0/src/printer.rs:105:17
  15:     0x560e5295e309 - <pretty_assertions::StrComparison<TLeft,TRight> as core::fmt::Display>::fmt::h8a9cbd32bbc3e26c
                               at /home/consulting/.cargo/registry/src/github.com-1ecc6299db9ec823/pretty_assertions-1.2.0/src/lib.rs:210:9
  16:     0x560e52baf38c - core::fmt::write::hdc75737b203416ad
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/fmt/mod.rs:1190:17
  17:     0x560e52b8bfec - core::fmt::Write::write_fmt::h4daf3e4aa36c80aa
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/fmt/mod.rs:186:9
  18:     0x560e52b8bfec - std::panicking::begin_panic_handler::PanicPayload::fill::{{closure}}::h982bfedc4aa80a74
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:550:22
  19:     0x560e52b8bfec - core::option::Option<T>::get_or_insert_with::h3554c0ea47abd828
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/option.rs:1522:49
  20:     0x560e52b8bfec - std::panicking::begin_panic_handler::PanicPayload::fill::h88ece18dee6cc20d
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:548:13
  21:     0x560e52b8bfec - <std::panicking::begin_panic_handler::PanicPayload as core::panic::BoxMeUp>::get::hae84bf37ecbba0f5
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:566:13
  22:     0x560e52b8c456 - std::panicking::rust_panic_with_hook::h0a743d387214ad02
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:697:34
  23:     0x560e52b8c157 - std::panicking::begin_panic_handler::{{closure}}::h394de780b6be0d14
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:588:13
  24:     0x560e52b89f14 - std::sys_common::backtrace::__rust_end_short_backtrace::h17d8cbafdb573507
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/sys_common/backtrace.rs:138:18
  25:     0x560e52b8be69 - rust_begin_unwind
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:584:5
  26:     0x560e52951863 - core::panicking::panic_fmt::h14d7282acf6a82b7
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/panicking.rs:143:14
  27:     0x560e5295f07d - codegen::single_function::h1ce943069e24d0cd
                               at /home/consulting/Documents/hotg-ai/wit-bindgen-wasm3/codegen/tests/codegen.rs:26:5
  28:     0x560e52954a6a - codegen::single_function::{{closure}}::h428a99483f218426
                               at /home/consulting/Documents/hotg-ai/wit-bindgen-wasm3/codegen/tests/codegen.rs:11:1
  29:     0x560e5295bc9e - core::ops::function::FnOnce::call_once::h1929ea5afa389f2e
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/ops/function.rs:227:5
  30:     0x560e5298acf3 - core::ops::function::FnOnce::call_once::hd7427b4f1fb1a3df
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/ops/function.rs:227:5
  31:     0x560e5298acf3 - test::__rust_begin_short_backtrace::hb863b53415a6c3c8
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/test/src/lib.rs:575:5
  32:     0x560e5298997f - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h6e7d1f9d5fd8a255
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/alloc/src/boxed.rs:1854:9
  33:     0x560e5298997f - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::he22973c0157abcad
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/panic/unwind_safe.rs:271:9
  34:     0x560e5298997f - std::panicking::try::do_call::hc40de4a913926a63
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:492:40
  35:     0x560e5298997f - std::panicking::try::h2ce93d83ee76f146
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:456:19
  36:     0x560e5298997f - std::panic::catch_unwind::hea9bd4ca39274840
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panic.rs:137:14
  37:     0x560e5298997f - test::run_test_in_process::h66e031868634c23a
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/test/src/lib.rs:598:18
  38:     0x560e5298997f - test::run_test::run_test_inner::{{closure}}::h374c2d96e09a6fdc
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/test/src/lib.rs:492:39
  39:     0x560e52993741 - test::run_test::run_test_inner::{{closure}}::h04c08617f7ade054
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/test/src/lib.rs:519:37
  40:     0x560e52993741 - std::sys_common::backtrace::__rust_begin_short_backtrace::h9ba773678c27c051
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/sys_common/backtrace.rs:122:18
  41:     0x560e5295f7df - std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}::hbe6a7761dbab7a63
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/thread/mod.rs:498:17
  42:     0x560e5295f7df - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h27450c879e029fc9
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/panic/unwind_safe.rs:271:9
  43:     0x560e5295f7df - std::panicking::try::do_call::h61592e92ca7f0dc1
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:492:40
  44:     0x560e5295f7df - std::panicking::try::h200a14ed0a35be11
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panicking.rs:456:19
  45:     0x560e5295f7df - std::panic::catch_unwind::hbfa0137f754138a7
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/panic.rs:137:14
  46:     0x560e5295f7df - std::thread::Builder::spawn_unchecked_::{{closure}}::h5bb3256bb0dd33f1
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/thread/mod.rs:497:30
  47:     0x560e5295f7df - core::ops::function::FnOnce::call_once{{vtable.shim}}::hfa0492591cec1bda
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/core/src/ops/function.rs:227:5
  48:     0x560e52b922a3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h8ab0aa77c9fe664a
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/alloc/src/boxed.rs:1854:9
  49:     0x560e52b922a3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hd82178596e9e381e
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/alloc/src/boxed.rs:1854:9
  50:     0x560e52b922a3 - std::sys::unix::thread::Thread::new::thread_start::hf4d984fb21d4c83d
                               at /rustc/4ce3749235fc31d15ebd444b038a9877e8c700d7/library/std/src/sys/unix/thread.rs:108:17
  51:     0x7ff4b471b5c2 - start_thread
  52:     0x7ff4b47a0584 - __clone
  53:                0x0 - <unknown>
thread panicked while panicking. aborting.
error: test failed, to rerun pass '--test codegen'

Caused by:
  process didn't exit successfully: `/home/consulting/Documents/hotg-ai/wit-bindgen-wasm3/target/debug/deps/codegen-e5f76d6e17dfe8c2 --nocapture` (signal: 6, SIGABRT: process abort signal)
```
</details>

The two strings I'm testing are the placeholder, `"asd"` on the right and the following generated Rust code on the left:

<details>
<summary>left hand side</summary>

```rust
pub mod single_function {
  #[allow(unused_imports)]
  use wit_bindgen_wasmtime::{wasmtime, anyhow};

  /// Auxiliary data associated with the wasm exports.
  ///
  /// This is required to be stored within the data of a
  /// `Store<T>` itself so lifting/lowering state can be managed
  /// when translating between the host and wasm.
  #[derive(Default)]
  pub struct SingleFunctionData {
  }
  pub struct SingleFunction<T> {
    get_state: Box<dyn Fn(&mut T) -> &mut SingleFunctionData + Send + Sync>,
    f: wasmtime::TypedFunc<(), ()>,
  }
  impl<T> SingleFunction<T> {
    #[allow(unused_variables)]

    /// Adds any intrinsics, if necessary for this exported wasm
    /// functionality to the `linker` provided.
    ///
    /// The `get_state` closure is required to access the
    /// auxiliary data necessary for these wasm exports from
    /// the general store's state.
    pub fn add_to_linker(
    linker: &mut wasmtime::Linker<T>,
    get_state: impl Fn(&mut T) -> &mut SingleFunctionData + Send + Sync + Copy + 'static,
    ) -> anyhow::Result<()> {
      Ok(())
    }

    /// Instantiates the provided `module` using the specified
    /// parameters, wrapping up the result in a structure that
    /// translates between wasm and the host.
    ///
    /// The `linker` provided will have intrinsics added to it
    /// automatically, so it's not necessary to call
    /// `add_to_linker` beforehand. This function will
    /// instantiate the `module` otherwise using `linker`, and
    /// both an instance of this structure and the underlying
    /// `wasmtime::Instance` will be returned.
    ///
    /// The `get_state` parameter is used to access the
    /// auxiliary state necessary for these wasm exports from
    /// the general store state `T`.
    pub fn instantiate(
    mut store: impl wasmtime::AsContextMut<Data = T>,
    module: &wasmtime::Module,
    linker: &mut wasmtime::Linker<T>,
    get_state: impl Fn(&mut T) -> &mut SingleFunctionData + Send + Sync + Copy + 'static,
    ) -> anyhow::Result<(Self, wasmtime::Instance)> {
      Self::add_to_linker(linker, get_state)?;
      let instance = linker.instantiate(&mut store, module)?;
      Ok((Self::new(store, &instance,get_state)?, instance))
    }

    /// Low-level creation wrapper for wrapping up the exports
    /// of the `instance` provided in this structure of wasm
    /// exports.
    ///
    /// This function will extract exports from the `instance`
    /// defined within `store` and wrap them all up in the
    /// returned structure which can be used to interact with
    /// the wasm module.
    pub fn new(
    mut store: impl wasmtime::AsContextMut<Data = T>,
    instance: &wasmtime::Instance,
    get_state: impl Fn(&mut T) -> &mut SingleFunctionData + Send + Sync + Copy + 'static,
    ) -> anyhow::Result<Self> {
      let mut store = store.as_context_mut();
      let f= instance.get_typed_func::<(), (), _>(&mut store, "f")?;
      Ok(SingleFunction{
        f,
        get_state: Box::new(get_state),

      })
    }
    pub fn f(&self, mut caller: impl wasmtime::AsContextMut<Data = T>,)-> Result<(), wasmtime::Trap> {
      self.f.call(&mut caller, ())?;
      Ok(())
    }
  }
}
```
</details>

I'm struggling to make a minimal example which still triggers the double-panic, so you can find the full code in the `codegen` directory of https://github.com/hotg-ai/wit-bindgen-wasm3/tree/e415bbfdbf73dc2f91b5cfff9e81b85903c4f566.

**Comment by tommilligan** [comment]
Hi - sorry for the slow response. Thank you for reporting this issue, it's definitely a bug and I can confirm I have a minimal reproduction of it.

The issue occurs when the left hand side of the diff contains a trailing newline, and the right hand side does not. As a temporary workaround, you can either:

- update your placeholder value to include a trailing newline, `"asd\n"`
- call `value.trim_end()` on both values before asserting, if you don't care about whitespace

_A_ fix for this is easy, I'm just trying to work out what the right behaviour should be internally!

**Comment by tommilligan** [comment]
A fix has been released in `1.2.1` - please let me know if it doesn't resolve your original case. Thanks again for reporting the issue!

**Comment by Michael-F-Bryan** [comment]
Thanks @tommilligan!
